sycl/include/CL/sycl/detail/image_accessor_util.hpp

```diff
@@ -144,67 +257,431 @@ void writePixel(vec<T, 4> Pixel, T *Ptr, image_channel_order Order) {
+  switch (ImageChannelType) {
+  case image_channel_type::unsigned_int8:
+    // convert_uchar_sat(Data)
+    Temp = (cl::sycl::clamp(WriteData, (uint)min_v<uint8_t>(),
+                            (uint)max_v<uint8_t>()));
+    PixelData = Temp.convert<ChannelType>();
+    break;
+  case image_channel_type::unsigned_int16:
+    // convert_ushort_sat(Data)
+    Temp = cl::sycl::clamp(WriteData, (uint)min_v<uint16_t>(),
+                           (uint)max_v<uint16_t>());
+    PixelData = Temp.convert<ChannelType>();
+    break;
+  case image_channel_type::unsigned_int32:
+    // no conversion is performed.
+    PixelData = WriteData.convert<ChannelType>();
+    break;
+  default:
+    // OpenCL Spec section 6.12.14.4 does not allow writing uint4 data to an
+    // image with channel datatype other than unsigned_int8,unsigned_int16 and
+    // unsigned_int32.
+    assert(!"Datatype of data to write - cl_uint4 is incompatible with the "
+            "image_channel_type of the image.");
+    break;
+  }
   return PixelData;
 }
 
-// Separate function for half datatype.
-// To resolve ambiguity when converting data into half datatype,float is used as
-// typecast.
-template <typename WriteDataType>
-vec<cl_half, 4> convertWriteDataToHalf(vec<WriteDataType, 4> WriteData,
-                                       image_channel_type ImageChannelType) {
-  vec<cl_half, 4> PixelData;
-  PixelData.x() = (float)WriteData.x();
-  PixelData.y() = (float)WriteData.y();
-  PixelData.z() = (float)WriteData.z();
-  PixelData.w() = (float)WriteData.w();
+template <typename ChannelType>
+vec<ChannelType, 4> convertWriteData(vec<cl_int, 4> WriteData,
+                                     image_channel_type ImageChannelType) {
+
+  vec<ChannelType, 4> PixelData;
+  vec<cl_int, 4> Temp;
+  switch (ImageChannelType) {
+  case image_channel_type::signed_int8:
+    // convert_char_sat(Data)
+    Temp =
+        cl::sycl::clamp(WriteData, (int)min_v<int8_t>(), (int)max_v<int8_t>());
+    PixelData = Temp.convert<ChannelType>();
+    break;
+  case image_channel_type::signed_int16:
+    // convert_short_sat(Data)
+    Temp = cl::sycl::clamp(WriteData, (int)min_v<int16_t>(),
+                           (int)max_v<int16_t>());
+    PixelData = Temp.convert<ChannelType>();
+    break;
+  case image_channel_type::signed_int32:
+    PixelData = WriteData.convert<ChannelType>();
+    break;
+  default:
+    // OpenCL Spec section 6.12.14.4 does not allow writing int4 data to an
+    // image with channel datatype other than signed_int8,signed_int16 and
+    // signed_int32.
+    assert(!"Datatype of data to write - cl_int4 is incompatible with the "
+            "image_channel_type of the image.");
+    break;
+  }
   return PixelData;
 }
 
+// Multiplication operator is not supported in vec class for a float operand.
+// So here each vector element is individually multiplied by the multiplicand.
+#define MULTIPLY_EACH(Result, Operand, Multiplicand)                           \
```

Comment:
Suggest converting it to function.

Reply:
I added it.
Any specific reason for suggesting a function usage here instead of a macro.